rust/unit_testable_rust_canister/src/hello_canister/tests/integration_tests.rs

```diff
@@ -0,0 +1,498 @@
+            wasm_path, e
+        )
+    })
+}
+
+/// Sets up NNS Governance canister with correct mainnet ID and production-like initialization
+/// Uses default_governance_for_tests() to get realistic initialization parameters
+fn setup_nns_governance(pic: &PocketIc) -> Principal {
+    let governance_wasm = get_governance_wasm();
+
+    // Parse the mainnet canister IDs
+    let nns_root_canister_id =
+        Principal::from_text(NNS_ROOT_CANISTER_ID).expect("Invalid NNS root canister ID");
+    let governance_canister_id = Principal::from_text(NNS_GOVERNANCE_CANISTER_ID)
+        .expect("Invalid NNS governance canister ID");
+
+    // Create the canister with the specific mainnet ID and NNS root as controller
+    pic.create_canister_with_id(Some(nns_root_canister_id), None, governance_canister_id)
+        .expect("Failed to create governance canister with specific ID");
+
+    // Use the governance types directly from our types.rs
+    let init_args = default_governance_for_tests();
+
+    println!("Installing NNS Governance canister with production-like configuration");
+
+    // Install the canister with the initialization arguments
+    pic.install_canister(
+        governance_canister_id,
+        governance_wasm,
+        encode_one(init_args).expect("Failed to encode governance init args"),
+        Some(nns_root_canister_id),
+    );
+
+    println!(
+        "NNS Governance canister setup complete at {}",
+        governance_canister_id
+    );
+
+    governance_canister_id
+}
+
+/// Test the counter functionality with Request/Response pattern
+#[test]
+fn test_counter_functionality() {
+    let pic = setup_pocket_ic();
+    let canister_id = deploy_hello_canister(&pic);
+
+    // Initial counter should be 0
+    let request = GetCountRequest {};
+    let response: GetCountResponse =
+        query(&pic, canister_id, "get_count", encode_one(request).unwrap());
+    assert_eq!(response.count, Some(0));
+
+    // Increment counter
+    let request = IncrementCountRequest {};
+    let response: IncrementCountResponse = update(
+        &pic,
+        canister_id,
+        "increment_count",
+        encode_one(request).unwrap(),
+    );
+    assert_eq!(response.new_count, Some(1));
+
+    // Check counter again
+    let request = GetCountRequest {};
+    let response: GetCountResponse =
+        query(&pic, canister_id, "get_count", encode_one(request).unwrap());
+    assert_eq!(response.count, Some(1));
+
+    // Increment again
+    let request = IncrementCountRequest {};
+    let response: IncrementCountResponse = update(
+        &pic,
+        canister_id,
+        "increment_count",
+        encode_one(request).unwrap(),
+    );
+    assert_eq!(response.new_count, Some(2));
+
+    // Final check
+    let request = GetCountRequest {};
+    let response: GetCountResponse =
+        query(&pic, canister_id, "get_count", encode_one(request).unwrap());
+    assert_eq!(response.count, Some(2));
+
+    // Increment counter
```

Comment:
Change to decrement, and fix args (types serialize the same way, so issue went unnoticed)